### PR TITLE
[WIP] Fixes for solidity and web3 tests

### PR DESCRIPTION
**Pull request by szotov**

### CODE REVIEW

File: nkms/blockchain/eth/interfaces.py [hunk 1]
```diff
@@ -220,6 +220,8 @@ def deploy_contract(self, contract_name: str, *args, **kwargs) -> Tuple[Contract
 
         deploy_transaction = {'from': self.deployer_address, 'gas': 2_400_000}
         deploy_bytecode = contract_factory.constructor(*args, **kwargs).buildTransaction(deploy_transaction)
+        if len(deploy_bytecode['data']) > 1000:
+            print("!!! " + contract_name + " : " + str(len(deploy_bytecode['data'])))
```
[the remark this answers is not in the document]
Comment (szotov) [reply nkms/blockchain/eth/interfaces.py]: My mistake, it's only for debug, I'll remove this

[the remark this answers is not in the document]
Comment (KPrasch) [reply nkms/blockchain/eth/interfaces.py]: Nonetheless - is bytecode size an issue?

File: tests/blockchain/eth/contracts/test_government.py [hunk 2]
```diff
@@ -233,7 +233,7 @@ def test_upgrade(web3, chain, escrow, policy_manager):
     node1 = web3.eth.accounts[1]
 
     # Deploy contract
-    government_library_v1, _ = chain.provider.get_or_deploy_contract(
+    government_library_v1, _ = chain.provider.deploy_contract(
```
[the remark this answers is not in the document]
Comment (szotov) [reply tests/blockchain/eth/contracts/test_government.py]: I completely agree with this decision, the tests will become more readable

[the remark this answers is not in the document]
Comment (KPrasch) [reply tests/blockchain/eth/contracts/test_government.py]: :+1: Awesome - this seems like an anti-pattern outside of testing, in any case.

File: nkms/blockchain/eth/sol/compile.py [hunk 3]
```diff
@@ -48,7 +48,7 @@ def compile(self) -> dict:
             sol_contract_paths.extend(glob.iglob(source_dir + '/**/*.sol', recursive=True))
 
         remapping_dirs = ["contracts={}".format(self._contract_source_dirs[0])]
-        compiled_sol = compile_files(sol_contract_paths, import_remappings=remapping_dirs)
+        compiled_sol = compile_files(sol_contract_paths, import_remappings=remapping_dirs, optimize=True)
```
Comment (KPrasch) [inline nkms/blockchain/eth/sol/compile.py]: What effect does optimize have on the compiler (Does this relate to bytecode size?)?

File: nkms/blockchain/eth/interfaces.py [hunk 4]
```diff
@@ -141,12 +141,12 @@ def dump_contract(self, address: str=None) -> dict:
         If no contract is found, it will raise NoKnownContract.
         """
         chain_data = self.dump_chain()
-        try:
-            contract_data = chain_data[address]
-        except KeyError:
-            raise self.UnknownContract('No known contract with address {}'.format(address))
-        else:
-            return contract_data
+        if address in chain_data:
```
Comment (KPrasch) [inline nkms/blockchain/eth/interfaces.py]: @tux and I deliberated about how to store this data - Does it make sense to have addresses as dict keys? What happens in the case of multiple deployments?

File: nkms/blockchain/eth/sol/compile.py (hunk 3 above)
Comment (szotov) [reply nkms/blockchain/eth/sol/compile.py]: First of all it reduces size of bytecode but we should find suitable number of steps (200 by default). Also it should optimize bytecode to using less gas

File: nkms/blockchain/eth/interfaces.py (hunk 4 above)
Comment (szotov) [reply nkms/blockchain/eth/interfaces.py]: For me it is not obvious how to store contracts. This implementation is close to populus. Also populus stores bytecode, but I'm not sure that will help much. When we will deploying multiple same contracts (Dispatcher, for example) then we can't find out wich contract should be returned using only contract name

File: nkms/blockchain/eth/interfaces.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (szotov) [reply nkms/blockchain/eth/interfaces.py]: Problem with Contract_v3 and test_dispatcher. Decreasing size of contracts helped to run test

**Comment by KPrasch** [comment]
I'm inclined to merge this even though it is still WIP.

**Comment by KPrasch** [comment]
Included upstream PR #251

**Pull request closed without merging**